packages/hardhat-core/src/internal/hardhat-network/jsonrpc/client.ts

```diff
@@ -309,8 +320,108 @@ export class JsonRpcClient {
       err.message.includes("header not found")
     );
   }
-}
 
-function blockTagToString(blockTag: BlockTag) {
-  return BN.isBN(blockTag) ? `0x${blockTag.toString("hex")}` : blockTag;
+  private _getCacheKey(method: string, params: any[]) {
+    const networkId = this.getNetworkId();
+    const plaintextKey = `${networkId} ${method} ${JSON.stringify(params)}`;
+
+    const hashed = createNonCryptographicHashBasedIdentifier(
+      Buffer.from(plaintextKey, "utf8")
+    );
+
+    return hashed.toString("hex");
+  }
+
+  private _getBatchCacheKey(batch: Array<{ method: string; params: any[] }>) {
+    let fakeMethod = "";
+    const fakeParams = [];
+
+    for (const entry of batch) {
+      fakeMethod += entry.method;
+      fakeParams.push(...entry.params);
+    }
+
+    return this._getCacheKey(fakeMethod, fakeParams);
+  }
+
+  private _getFromCache(cacheKey: string): any | undefined {
+    return this._cache.get(cacheKey);
+  }
+
+  private _storeInCache(cacheKey: string, decodedResult: any) {
+    this._cache.set(cacheKey, decodedResult);
+  }
+
+  private async _getFromDiskCache(
+    cacheKey: string,
+    tType: t.Type<any>
+  ): Promise<any | undefined> {
+    const rawResult = await this._getRawFromDiskCache(cacheKey);
+
+    if (rawResult !== undefined) {
+      return decode(rawResult, tType);
+    }
+  }
+
+  private async _getBatchFromDiskCache(
+    cacheKey: string,
+    tTypes: Array<t.Type<any>>
+  ): Promise<any[] | undefined> {
+    const rawResults = await this._getRawFromDiskCache(cacheKey);
+
+    if (!Array.isArray(rawResults)) {
+      return undefined;
+    }
+
+    return rawResults.map((r, i) => decode(r, tTypes[i]));
+  }
+
+  private async _getRawFromDiskCache(
+    cacheKey: string
+  ): Promise<any | undefined> {
+    try {
+      return await fsExtra.readJSON(this._getDiskCachePathForKey(cacheKey), {
+        encoding: "utf8",
+      });
+    } catch (error) {
+      return undefined;
```

Comment:
Maybe we should add a debug log here?

Reply:
That was pretty sloppy. It should only return undefined if the error is because the file doesn't exist.